helix-term/src/commands/typed.rs

```diff
@@ -5,19 +5,45 @@ use crate::job::Job;
 
 use super::*;
 
-use helix_core::encoding;
+use helix_core::{encoding, shellwords::Shellwords};
 use helix_view::editor::{Action, CloseError, ConfigEvent};
 use serde_json::Value;
 use ui::completers::{self, Completer};
 
 #[derive(Clone)]
-pub struct TypableCommand {
+pub struct TypableCommand<'a> {
     pub name: &'static str,
     pub aliases: &'static [&'static str],
     pub doc: &'static str,
     // params, flags, helper, completer
     pub fun: fn(&mut compositor::Context, &[Cow<str>], PromptEvent) -> anyhow::Result<()>,
-    pub completer: Option<Completer>,
+    /// What completion methods, if any, does this command have?
+    pub completer: CompletionConfig<'a>,
+}
+
+#[derive(Clone)]
+pub enum CompletionConfig<'a> {
```

Comment:
I don't think an enum is an amazing fit here. NoComoletions is just a subset of positional and this approach doesn't allow commands which have a couple positional arguments followed by varags. Instead you should simply use a list of positional arguments followed by an optional vararg completed:

``` rust
strict CommandSignature{
   postional_args: &'static [Option<Completer>],
   var_arcs: Option<Completer>
} 
```

You can get the compler for field n with:
``` rust
self.positional_arts.get(n).unwrap_or(self.var_args)
```

Note that my suggestion uses a `'ststic` slice. Tgis removes the lifetime argument you were unaopy with and is a good fit here because currently all tupable commands are static anyway and if we move away from that we need to heap allocate anyway for that cade a cow would be used)